### ISSUE TITLE
Plugin state is lost on hydration

### ISSUE THREAD

**Issue by scriptcoded** [nuxt/nuxt.js#8852]
### Versions

- nuxt: 2.15.0
- node: v14.15.2

### Reproduction

https://codesandbox.io/s/confident-johnson-gghmq?file=/pages/index.vue

The only relevant files are `pages/index.vue`, `plugins/stateTest.js` and `nuxt.config.js` (to include the plugin).

### Steps to reproduce

1. Write a plugin that has some injected state
2. Modify the state inside `fetch`
3. Observe that the state is not updated on the client side.
4. Observe that the state **is** updated in the server side rendered output (view-source:https://gghmq.sse.codesandbox.io/).

### What is Expected?

The client state should match the server side state

### What is actually happening?

The client state is never updated, while the server side state is.

---

I encountered this issue while creating a plugin for managing state from the server. By doing `await this.$myPlugin.users.load()` and then accessing that data using `this.$myPlugin.users.items` I could greatly simplify both templates and data loading. While moving this to the `fetch` hook however, It didn't work. I found out that this was the root cause.

I'm not sure if this is necessarily a bug, but if nothing else it seems like something that might've been overlooked (or I've missed some important documentation).

**Comment by danielroe** [comment]
@scriptcoded As you're creating a custom piece of state, Nuxt can't know to hydrate it as you want. (It could, for example, contain methods or things that shouldn't or couldn't be hydrated.) Instead, you can manage it yourself using [`beforeNuxtRender`](https://nuxtjs.org/docs/2.x/internals-glossary/context#beforenuxtrender) and [`nuxtState`](https://nuxtjs.org/docs/2.x/internals-glossary/context#nuxtstate) - two handy properties available on the context.

Hope this helps :blush:

**Comment by scriptcoded** [comment]
@danielroe Thank you! I'll give that a shot :blush:

**Comment by scriptcoded** [comment]
@danielroe That worked flawlessly. Thank you so much, saved us a ton of time we'd spend refactoring otherwise ^_^

**Comment by scriptcoded** [comment]
For anybody having the same issues I've had in the future, the solution actually turned out to be a bit more convoluted.

In our case we had some asynchronous tasks being run within a few components. Since these were async it usually happened that `beforeNuxtRender` was called way before our tasks were finished. This caused a lot of issues with hydration.

The solution for us was to instead use a Nuxt module that provides a plugin. This plugin injects the state when doing SSR, and loads it when on the client:
```js
// Some state that might be modified at runtime
const myState = Vue.observable({})

if (process.server) {
  ctx.ssrContext.__myState = myState
} else {
  myState = ctx.nuxtState.__myState
}

inject('myState', myState)
```

Though this isn't enough, as we still need to inject our state into `nuxtState`. This is where the module comes in. We listen on the `vue-renderer:ssr:context` hook and inject our state there:
```js
this.nuxt.hook('vue-renderer:ssr:context', (ctx: any) => {
  ctx.nuxt.__myState = ctx.__myState
})
```

So that solves the issue with injecting the state after the async tasks are finished. We did however have another issue which had to do with Nuxt trying to serialize our state. Because we had some data that Nuxt couldn't easily serialize we got a bunch of unclear "RangeError: Maximum call stack size exceeded" from `devalue.js`. A simple solution to this problem was to inject our state as a string:

In the plugin:
```js
myState = JSON.parse(ctx.nuxtState.__myState)
```

And the module:
```js
ctx.nuxt.__myState = JSON.stringify(ctx.__myState)```

Hope this helps someone, as we've been stuck with these issues for a month now :sweat_smile:
